### ISSUE TITLE
ExceptionMapper not handling Exceptions thrown from injected RestClient

### ISSUE THREAD

**Issue by irishshagua** [quarkusio/quarkus#4031]
**Describe the bug**
For an ExceptionMapper register with a controller, it is not being fired for exceptions which are thrown by an injected RestClient

**Expected behavior**
An `ExceptionMapper<Thowable>` should handle any exceptions which could be caught in the controller code

**Actual behavior**
`WebApplicationException`s which are thrown from a RestClient are not being handled by the `ExceptionMapper`

**To Reproduce**
There is a test showing this behavious in this [repo](https://github.com/irishshagua/quakus-exception-handler-issues)

**Additional context**
Initially raised as an issue on [StackOverflow](https://stackoverflow.com/questions/57910496/quarkus-exceptionmapper-does-not-handle-webapplicationexception)

**Comment by gsmet** [comment]
@kenfinnigan @asoldano could you have a look at that one? Not sure if the server or the client is the culprit. There is a reproducer.

**Comment by kenfinnigan** [comment]
I think the issue is that the REST Client doesn't have a media type set: https://github.com/irishshagua/quakus-exception-handler-issues/blob/master/src/main/java/demo/SomeRestClient.java#L13

Meaning it can't find a writer

**Comment by irishshagua** [comment]
> I think the issue is that the REST Client doesn't have a media type set: https://github.com/irishshagua/quakus-exception-handler-issues/blob/master/src/main/java/demo/SomeRestClient.java#L13
> 
> Meaning it can't find a writer

Hey @kenfinnigan ,
I tested again with a Produces and Consumes anno on the Rest client method and it didn't make any difference to the failing test case. Just to be clear, the actual REST client call in the test should fail. The unexpected behavior is that the ExceptionHandler is not being triggered.

**Comment by benneq** [comment]
I guess, I have the same issue, but with a different set of exceptions:

`ExceptionMapper<Exception>` (same applies to `ExceptionMapper<Throwable>`) won't catch `ResteasyViolationException`.

**But** with `ExceptionMapper<ResteasyViolationException>` (or `ExceptionMapper<ValidationException>`) it works.

Very weird behavior.

I'm using Quarkus 1.0.1

<details>
  <summary>pom.xml (Click to expand)</summary>

```
  <dependencies>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-junit5</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.rest-assured</groupId>
      <artifactId>rest-assured</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-jsonb</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-jwt</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-health</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-hibernate-validator</artifactId>
    </dependency>
  </dependencies>
```
</details>

<details>
  <summary>DefaultExceptionHandler class (Click to expand)</summary>

```
@Provider
public class DefaultExceptionHandler implements ExceptionMapper<Exception> {

	@Override
	public Response toResponse(Exception exception) {
		// doesn't work for ResteasyViolationException
		// won't get called
```
</details>

<details>
  <summary>ValidationExceptionHandler class (Click to expand)</summary>

```
@Provider
public class ValidationExceptionHandler implements ExceptionMapper<ResteasyViolationException> {

	@Override
	public Response toResponse(ResteasyViolationException exception) {
		// works as expected
```
</details>

<details>
  <summary>TestResource class (Click to expand)</summary>

```
@Path("/test")
public class TestResource {
    
    @POST
    public String test(@NotNull Boolean value) {
    	return "" + value;
    }
    
}
```
</details>

<details>
  <summary>Test Request JSON (Click to expand)</summary>

```
POST /test
```

Providing "no content" will trigger the `@NotNull` constraint validation error.
</details>

**Comment by kenfinnigan** [comment]
Is it possible to update to 1.8.1.Final and see if the error is still present?

**Comment by irishshagua** [comment]
> Is it possible to update to 1.8.1.Final and see if the error is still present?

I bumped the [reproducable](https://github.com/irishshagua/quakus-exception-handler-issues/commit/1ef45e030daf207c98696c49ac17a6af9f84c0b1) to 1.8.1.Final. Test is still failing.

**Comment by kenfinnigan** [comment]
Taking another look at the example code and the spec, I think the issue is that the MP REST Client spec does not support exception mappers marked with `@Provider` unless `@RegisterProvider` as on the interface to explicitly register the exception mapper with the client.

Take a look at https://download.eclipse.org/microprofile/microprofile-rest-client-1.4.1/microprofile-rest-client-1.4.1.html#_responseexceptionmapper on the specific exception mapper handling for MP REST Client

**Comment by pazkooda** [comment]
## Works as expected
... most likely i.e assuming that `WebApplicationException` thrown by `RestClient` contains response body

## Explanation

### TL;DR:

As of [JaxRS specification](https://raw.githubusercontent.com/javaee/jax-rs-spec/master/spec.pdf) _Section 3.3.4_ exception mappers shouldn't not process `WebApplicationException` if such one contains response with entity(/body) and serve that response itself.

### Longer read

[JaxRS specification](https://raw.githubusercontent.com/javaee/jax-rs-spec/master/spec.pdf) _Section 3.3.4_ _Point 1_ states:
> (...)
Instances of `WebApplicationException` and its subclasses MUST be mapped to a response as follows. If the response property of the exception does not contain an entity and an exception mapping
provider (see Section 4.4) is available for `WebApplicationException` or the corresponding subclass, an implementation MUST use the provider to create a new Response instance, otherwise the
`response` property is used directly. 
(...)

and RestEasy implements it with [this line of code](https://github.com/resteasy/Resteasy/blob/0ef00121d11f8ca67af6c2549419fb5fc36fe391/resteasy-core/src/main/java/org/jboss/resteasy/core/ExceptionHandler.java#L339).

As I assume many of you I consider this also a bit unfortunate - I'd expect `ExceptionMapper` to be triggered nevertheless. Yet ReasEasy is here to provide implementation of spec - not to make it work, the what seems to be, most natural way.

## Solution

Use `@RegisterProvider` in your RestClient as most likely this is what Jax-RS spec is requiring you to do.

**Comment by kenfinnigan** [comment]
Closing this as I don't believe it's an issue with Quarkus, please re-open if that's not the case

**Comment by pazkooda** [comment]
For those who are interested in related (good❕) changes in this area in Quarkus 1.12 I strongly advice following discussion in related #15272 starting from [this comment](https://github.com/quarkusio/quarkus/issues/15272#issuecomment-799775600)

**Comment by anduda** [comment]
I had WebApplicationException from external dependenies(Keycloak) and needed to map response my error response model.
You can handle it by implementing ContainerResponseFilter and do the same checks as in Quarkus errorHandler. Something like this:

`
@Provider
public class WebExceptionMapper implements ContainerResponseFilter {

    @Override
    public void filter(
            ContainerRequestContext requestContext,
            ContainerResponseContext responseContext
    ) {
        if (requestContext instanceof ResteasyReactiveContainerRequestContext) {
            var resteasyRequestContext = (ResteasyReactiveContainerRequestContext) requestContext;
            if (resteasyRequestContext.getServerRequestContext() instanceof QuarkusResteasyReactiveRequestContext) {
                var quarkusRequestContext = (QuarkusResteasyReactiveRequestContext) resteasyRequestContext
                        .getServerRequestContext();
                var throwable = quarkusRequestContext.getThrowable();
                if (throwable instanceof WebApplicationException) {
                    //map exception to your error response structure
                    var errorResponse = mapResponse((WebApplicationException) throwable);
                    quarkusRequestContext.setResult(errorResponse);
                }
            }
        }
    }

}
`
